### PR TITLE
Added web3 start command

### PR DESCRIPTION
**Pull request by treeder**
* Starts a local GoChain dev node using Docker.
* Funds private key when starting a node
* Adds `web3 account create` command - part of #51

### CODE REVIEW

**Comment by jmank88** [comment]
It might be simpler (and faster) to have web3 manage the accounts itself and ask the node to fund them. It could even recognize that WEB_PRIVATE_KEY is set, and automatically fund that account.

**Comment by treeder** [comment]
Is that maybe a different issue?  This is just a short hand for running the docker command.

**Comment by jmank88** [comment]
> Is that maybe a different issue? This is just a short hand for running the docker command.

To be useful, it would have to parse the output from the docker logs for the user to reenter. I'm proposing that be avoided by setting the `--local.fund` flag instead, which can take already existing accounts or newly created ones.

**Comment by treeder** [comment]
Are you saying the docker command that's currently in the README isn't useful either? This does the exact same thing.

**Comment by jmank88** [comment]
Running detached would hide info. The readme isn't detached so the relevant info shows. Using `local.fund` makes this a non-issue though, and allows for other tricks.

**Comment by treeder** [comment]
This doesn't run it detached, where do you see that?

**Comment by jmank88** [comment]
> This doesn't run it detached, where do you see that? It's an option, but not default.

Why provide an option that they can't use? Regardless, why not provide them a better option in the first place? Using `local.fund` is a better user experience, and allows detaching to be the default: they don't have to find new keys in the logs to set flags or env vars; the ones they are already using can be directly funded, or new accounts can be created on the fly and logged plainly and w/o other noise (no extracting them from docker log output).

File: cmd/web3/start.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		log.Fatalln("Starting command failed:", err)
+	}
+
+	done := make(chan error, 1)
+	go func() {
+		done <- cmd.Wait()
+	}()
+	// catch ctrl-c and kill
+	sigC := make(chan os.Signal, 2)
+	signal.Notify(sigC, os.Interrupt, syscall.SIGTERM)
```
Comment (jmank88) [inline cmd/web3/start.go]: Signal interrupt cancellation is already handled in `main`. You should be able to just use `exec.CommandContext` directly without any extra fuss.

**Comment by treeder** [comment]
> Regardless, why not provide them a better option in the first place? Using `local.fund` is a better user experience, and allows detaching to be the default: they don't have to find new keys in the logs to set flags or env vars; the ones they are already using can be directly funded, or new accounts can be created on the fly and logged plainly and w/o other noise (no extracting them from docker log output).

Can you explain step by step how this option would work from a user perspective?

**Comment by jmank88** [comment]
## Current user perspective:
### No `WEB3_PRIVATE_KEY` + attached:
After starting, user must copy keys from logs to use. They will probably ignore the rest of the logs.
### No `WEB3_PRIVATE_KEY` + detached: 
They must additionally open the docker logs first, then copy the keys.

### With `WEB3_PRIVATE_KEY` set:
Doesn't matter, they must still get keys from the logs first.

## Proposal using `local.fund`:
Docker is always detached (but still available with full logs).
### No `WEB3_PRIVATE_KEY`:
Keys are logged cleanly to the console in whatever format we want. No other logs are in the way.
### With `WEB3_PRIVATE_KEY` set:
Account is automatically funded, and user can immediately begin running signing commands. Other accounts could optionally be created or not - this becomes a web3 level option, instead of a gochain level option.

**Comment by treeder** [comment]
So step by step, something like this:

```
web3 create account
# copy private key returned
export WEB3_PRIVATE_KEY=X
web3 start --local.fund
```

Something like that?

**Comment by jmank88** [comment]
Yes, basically. The user wouldn't have to pass any extra options to `web3 start`. `WEB3_PRIVATE_KEY` can be automatically noticed and funded transparently behind the scenes. If it's not there, then web3 creates an account and still uses `local.fund`, and the user gets the private key to set afterwards. Both paths are faster and simpler than letting `gochain` create the accounts itself.

**Comment by treeder** [comment]
Ok, sounds fine. So fund the current WEB3_PRIVATE_KEY + create 9 more accounts, pass those into the docker run command to get funded then print them to the console. 
If no WEB3_PRIVATE_KEY, make 10 accounts.

**Comment by jmank88** [comment]
Or even just fund a single account by default for speed, and add a flag for requesting more funded accounts. Related, I created: https://github.com/gochain-io/gochain/issues/360

File: cmd/web3/start.go [hunk 2]
```diff
@@ -57,6 +75,7 @@ func start(c *cli.Context) error {
 			args = append(args, "-d")
 		}
 		args = append(args, "gochain/gochain", "--local")
+		args = append(args, "--local.fund")
```
Comment (jmank88) [inline cmd/web3/start.go]: Need to pass the account here

File: cmd/web3/start.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	"os/signal"
+	"syscall"
+
+	"github.com/gochain-io/web3"
+	"github.com/urfave/cli"
+)
+
+func start(c *cli.Context) error {
+
+	privateKey := os.Getenv("WEB3_PRIVATE_KEY")
```
Comment (jmank88) [inline cmd/web3/start.go]: The other cases pull this through the cli package API as a flag with corresponding env var

File: accounts.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+}
+func ParsePrivateKey(pk string) (*Account, error) {
+
+	key, err := crypto.GenerateKey()
+	if err != nil {
+		return nil, err
+	}
+	return &Account{
+		key: key,
+	}, nil
```
Comment (jmank88) [inline accounts.go]: Copy/paste template leftover here?

**Comment by treeder** [comment]
wip, don't look yet. :)  just switching computers.

**Comment by treeder** [comment]
Alright, now it all works.

File: cmd/web3/main.go [hunk 5]
```diff
@@ -215,6 +226,68 @@ func main() {
...
+				if pk == "" {
+					fmt.Printf("%v not set", pkVarName)
+					return
+				}
+				acc, err := web3.ParsePrivateKey(pk)
+				if err != nil {
+					log.Fatalln(err)
+				}
+				fmt.Print(acc.PublicKey())
+			},
```
Comment (jmank88) [inline cmd/web3/main.go]: What if this was under `account` (no subcommand) instead?
`web3 account` and `web3 account create`

File: accounts.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+}
+func ParsePrivateKey(pkHex string) (*Account, error) {
+	fromPK := Strip0x(pkHex)
```
Comment (jmank88) [inline accounts.go]:
```suggestion
	fromPK := strings.TrimPrefix(pkHex,"0x")
```

File: cmd/web3/start.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+		}
+		args = append(args, "gochain/gochain", "--local")
+		args = append(args, "--local.fund", acc.PublicKey())
+		cmd = exec.Command("docker", args...)
+		fmt.Println("Starting your own, personal GoChain instance...")
+		fmt.Println(asciiLogo)
+		fmt.Println()
+		if privateKey == "" {
+			fmt.Printf("We created an account for you to get started quickly.\n\nYour private key is:\n\n%v\n\n"+
+				"Type: `export WEB3_PRIVATE_KEY=%v` to get make using this tool easier.\n\n", acc.PrivateKey(), acc.PrivateKey())
```
Comment (jmank88) [inline cmd/web3/start.go]:
```suggestion
				"Type: `export WEB3_PRIVATE_KEY=%v` to make using this tool easier.\n\n", acc.PrivateKey(), acc.PrivateKey())
```

File: cmd/web3/start.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+		}
+		fmt.Printf("You're account %v is pre-funded with %v GO.\n", acc.PublicKey(), 1000)
```
Comment (jmank88) [inline cmd/web3/start.go]:
```suggestion
		fmt.Printf("Your account %v is pre-funded with %v GO.\n", acc.PublicKey(), 1000)
```

File: cmd/web3/main.go (hunk 5 above)
Comment (treeder) [reply cmd/web3/main.go]: Hmm, would there every be more to account than just the raw address?  That said, maybe `web3 account address` or something would be appropriate.

**Pull request merged**